trl/trainer/srpo_trainer.py

```diff
@@ -0,0 +1,1755 @@
+            metrics[f"{prefix}ratios/{key}"] = item.detach().mean().cpu()
+        if self.args.rpo_alpha is not None:
+            metrics[f"{prefix}nll_loss"] = policy_nll_loss.detach().mean().cpu()
+
+        return losses.mean(), metrics
+
+    def compute_loss(
+        self,
+        model: Union[PreTrainedModel, nn.Module],
+        inputs: Dict[str, Union[torch.Tensor, Any]],
+        return_outputs=False,
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, Dict[str, torch.Tensor]]]:
+        if not self.use_dpo_data_collator:
+            warnings.warn(
+                "compute_loss is only implemented for DPODataCollatorWithPadding, and you passed a datacollator that is different than "
+                "DPODataCollatorWithPadding - you might see unexpected behavior. Alternatively, you can implement your own prediction_step method if you are using a custom data collator"
+            )
+
+        compute_loss_context_manager = (
+            torch.cuda.amp.autocast
+            if self._peft_has_been_casted_to_bf16
+            else nullcontext
+        )
+
+        with compute_loss_context_manager():
+            loss, metrics = self.get_batch_loss_metrics(
+                model, inputs, train_eval="train"
+            )
+
+        # Make sure to move the loss to the device the original accumulating loss is at back in the `Trainer` class:
+        loss = loss.to(self.args.device)
+        # force log the metrics
+        self.store_metrics(metrics, train_eval="train")
+
+        if return_outputs:
+            return (loss, metrics)
+        return loss
+
+    def get_batch_samples(
+        self, model, batch: Dict[str, torch.LongTensor]
+    ) -> Tuple[List[str], List[str], List[List[str]]]:
+        """Generate samples from the model and reference model for the given batch of inputs."""
+
+        # If one uses `generate_during_eval` with peft + bf16, we need to explicitly call generate with
+        # the torch cuda amp context manager as some hidden states are silently casted to full precision.
+        generate_context_manager = (
+            nullcontext
+            if not self._peft_has_been_casted_to_bf16
+            else torch.cuda.amp.autocast
+        )
+
+        with generate_context_manager():
+            policy_output = model.generate(
+                input_ids=batch["zero_prompt_prompt_input_ids"],
+                attention_mask=batch["zero_prompt_prompt_attention_mask"],
+                max_length=self.max_length,
+                do_sample=True,
+                pad_token_id=self.tokenizer.pad_token_id,
+            )
+            policy_output_decoded = self.tokenizer.batch_decode(
+                policy_output, skip_special_tokens=True
+            )
+            revisions = []
+            for i in range(len(policy_output_decoded)):
+                revisions.append([])
+            prev_prompts = batch["prompt"]
+            self.tokenizer.padding_side = "left"
+            for n in range(5):
```

Comment:
For eval we do 5 revision rollouts as seen in the paper. Might be worth making this configurable.